### ISSUE TITLE
JetStream does not redeliver until AckWait reached even on client disconnects

### ISSUE THREAD

**Issue by crosem** [nats-io/nats-server#4716]
### Observed behavior

With an async push subscriber on a durable JetStream consumer with a DeliveryGroup on a stream setup with RetentionPolicy.WorkQueue.  If a subscriber terminates before acking the msg that message is not redelivered until the AckWait period is reached.  Even if the connection to that subscriber is closed gracefully.

### Expected behavior

If a subscriber receives a message (pushed) and before being able to ack it does not and instead closes the subscription and terminates the connection I would expect that message to be re-delivered immediately instead of only after the Consumers AckWait period is reached.  I would hope the NATS server would be able to detect the disconnect and redeliver asap.

### Server and client version

Nats server 2.10.2 using C# NATS.Client 1.0.8 

### Host environment

Windows Server 2012

### Steps to reproduce

1) Create a work queue stream
2) Create a durable consumer on that stream with a deliver group, deliverysubject and an AckWait of 5 mins
3) Create a async push subscriber on that consumer
4) publish a message on that subject
5) Do not ack it and instead either terminate the process or shut it down cleanly (close sub and connection)
6) Start another process and wait for the un-acked message to come
7) It doesn't come until the 5 min AckWait period is done.

**Comment by derekcollison** [comment]
Clients self heal and reconnect on their own. The server is not trying to understand how the connection was closed and/or if the connection is coming back.

Working as designed.

**Comment by crosem** [comment]
Client re-connects is not the issue I am raising.  It's pending message
delivery after a disconnect for a subscriber of a consumer group.

Consider this real world scenario.  We have a trading system and a durable
consumer on a stream that is setup with a delivery group for horizontal
scaling.

A trade is sent and delivered to a subscriber in the delivery group. While
processing and before being able to ack the process crashes.  Since these
are time critical events we would hope that the nats server would recognize
the disconnect...meaning that subscriber in the delivery group will never
be able to ack the message so it should be re-delivered immediately to
another subscriber.

Instead we have to wait for the AckWait to expire. Now since these are
trades and can only be processed once we don't want to set the AckWait too
short, risking multiple bookings.  So we set it longer but now in this
scenario the trade is stuck in limbo waiting for AckWait and the trader
cannot see their updated trade.

**Comment by derekcollison** [comment]
Again, the server can not know that the app crashed and will not auto-reconnect almost instantly. All NATS clients can do this behavior.

So if it is time sensitive and you have a well defined nominal processing window you can set the ack timeout to that, or use NAK if the app can not handle, that will immediately redeliver the message to the next available.

**Comment by crosem** [comment]
Yes the client will reconnect immediately and I have other clients in that
delivery group available to receive the message but they won't get it until
the AckWait expires.

So as you say I need to tinker with AckWait so that I can achieve a
reasonable re-delivery time frame at the expense of possible harmful
duplication.

What my hope with NATS was and what I'm used to with other message brokers
(e.g. Tibco EMS) is that once the client disconnects any messages delivered
and pending an ack are re-delivered to other available consumers of the
subject immediately.

Basically NATS recognizing it will never receive an ACk from the subscriber
because the client is gone so why wait for the AckWait to expire?  Just
re-delivered it.  I would NAK it myself if I could but not possible in a
client that exits unexpectedly.

**Comment by derekcollison** [comment]
Again, the server does not know definitely that the client is gone forever.

And as FYI, I designed TIBCO's EMS system..

**Comment by crosem** [comment]
Oh wow...excellent product.  Used it for over 10 yrs..good job!

So comparing to Tibco EMS and it's design of point-to-point vs a similar
pattern in NATS with delivery groups and worker queue retention, what
differs in the NATS design that doesn't allow it to know when a client is
gone for good? Is it the consumer layer on top of the stream?

**Comment by derekcollison** [comment]
The fact that very few NATS systems are single servers, they were designed to be able to be plugged together in almost any topology. So the server housing the consumer leader would not be able to really understand the nature of the application unless it directly signaled, e.g. NAK. NATS does not require anything to close a connection, and that was by design, so no way of knowing the nature of the client application.

So for JetStream, we prefer to optimize for number of times a message is processed vs immediate re-delivery in the circumstance you provide above. We allow that behavior, but you have to send the server a positive signal. That is the NAK, and can travel over any random NATS system topology.
